### ISSUE TITLE
sokol_gfx: possible bug when using many instance buffers

### ISSUE THREAD

**Issue by floatms** [floooh/sokol#104]
Hi,

TL;DR: I used instanced rendering with about 400 streaming buffers per pass of 8 KB size each. Some instances do not get drawn when the count of buffers used is set to somewhere between 100 and 200. Buffer size doesn't seem to affect this. Using D3D backend and sokol_app.

I created a gist for this issue, see: [https://gist.github.com/floatms/d2466adf93cbb3126c8a655b6d08e275](https://gist.github.com/floatms/d2466adf93cbb3126c8a655b6d08e275)
Excuse the small formatting issues; don't know why github formats weirdly only some of the time.

I  was drawing instanced rectangles in my application using "primitive buffers" that supply all the instance data to the shader. It was all going great but I noticed that some of the rectangles that are drawn last where not showing up on screen during a stress test with about 100 000 instances (performance was still very good by the way, so kudos to you).
At first I thought there was an issue with my code as I had seen similar things before but I was able to reduce the issue and remove all non-sokol-specific code.

There's a comment with "BUG" prefixed in there that explains how to reproduce.
Set BUFFER_COUNT to (10) or even (100) and observe that there is a white and a red rectangle on screen.
Set BUFFER_COUNT to (400) and the red rectangle disappears.

The gist is kept close to the sokol-samples. It only depends on sgfx, sapp and HandmadeMath.

I hope you will be able to reproduce or even better find a mistake in my setup of sokol.

Cheers

**Comment by floooh** [comment]
Hi! Thanks for the detailed bug report, very useful. One thing that's a bit weird is that you're calling sg_begin_default_pass() several times per frame, usually there should only be one default-pass per frame, and that should be the last pass right before sg_commit(). I should probably check that in the validation layer. But at least in D3D11 this shouldn't be responsible for the problem you're seeing, it's just inefficient.

I'll try to have a closer look over the next few days.

One very useful tools for such problem is RenderDoc btw (https://renderdoc.org/), this let's you capture a frame and replay it step by step for each draw call, and inspect the internal D3D11 state. That's what I'll be using for debugging anyway, but just in case you want to have a look too ;)

Cheers!
-Floh.

**Comment by floooh** [comment]
Wait a minute... by default the pool size for buffers will be 128, see docs here: https://github.com/floooh/sokol/blob/60d9e935fd80cec97390a77d4315932fb07abaf1/sokol_gfx.h#L1070-L1075

So after creating 128 buffers, sg_create_buffer() will start to fail and return SG_INVALID_ID, and draw calls attempting to use those invalid buffer ids will fail.

However when compiling in debug mode, you should be getting failed assert- and validation-checks left and right (these checks are all disabled in release-mode).

Since in your gist you're not increasing the pool sizes, I suspect that this is what's happening: buffer creating starts failing after 128, but when compiling in release-mode, this would go unnoticed.

You can check the blend-sapp sample for an example where I need to increase a resource pool size (however in this case, for the pipeline resource pool, while in your case you'd need to increase the buffer_pool_size): https://github.com/floooh/sokol-samples/blob/059c8a662a66cef7ced54e29e949cbfc81d10cd1/sapp/blend-sapp.c#L41

Let me know if this fixes your problem (and you should definitely develop and test in debug mode, since this enables tons of validation checks) :)

**Comment by floooh** [comment]
This would be a "fixed" version of your frame function:

```c
void frame(void) {
    sg_begin_default_pass(&pass_action_clear, sapp_width(), sapp_height());
    for(int i = 0; i < BUFFER_COUNT; i++) {
        sg_update_buffer(buffers[i], rectangle_buffers[i], (MAX_BUFFER_SIZE / sizeof(rectangle_t)));
        bindings.vertex_buffers[1] = buffers[i];

        hmm_mat4 view = HMM_LookAt(HMM_Vec3(0.0f, 0.0f, 5.0f), HMM_Vec3(0.0f, 0.0f, 0.0f), HMM_Vec3(0.0f, 1.0f, 0.0f));
        hmm_mat4 proj = HMM_Orthographic(0.0f, sapp_width(), 0.0f, sapp_height(), -1.0f, 5.0f);
        hmm_mat4 view_proj = HMM_MultiplyMat4(proj, view);

        sg_apply_pipeline(rect_pipeline);
        sg_apply_bindings(&bindings);
        sg_apply_uniforms(SG_SHADERSTAGE_VS, 0, &view_proj, sizeof(hmm_mat4));
        sg_draw(0, 6, MAX_BUFFER_SIZE / sizeof(rectangle_t));
    }
    sg_end_pass();
    sg_commit();
}
```
The default-pass is for when you want to render to the standard-framebuffer that will be presented at the end of the frame, most simple 'rendering pipelines' only need a default pass. And there should only be one default-pass in the frame, and it should be the last pass you render.

sg_begin_pass() is used when you want to render to an offscreen 'render-target-image', which you want you want to use later in the frame as a texture-image.

I'm glad the issue is resolved, it's still valuable feedback for where I need to improve the documentation and/or validation layer.

Cheers!

**Comment by floooh** [comment]
...actually, this is a slightly better version of the frame function, note how the ```sg_apply_pipeline()``` has been moved out of the loop:

```c
void frame(void) {
    sg_begin_default_pass(&pass_action_clear, sapp_width(), sapp_height());
    sg_apply_pipeline(rect_pipeline);
    for(int i = 0; i < BUFFER_COUNT; i++) {
        sg_update_buffer(buffers[i], rectangle_buffers[i], (MAX_BUFFER_SIZE / sizeof(rectangle_t)));
        bindings.vertex_buffers[1] = buffers[i];

        hmm_mat4 view = HMM_LookAt(HMM_Vec3(0.0f, 0.0f, 5.0f), HMM_Vec3(0.0f, 0.0f, 0.0f), HMM_Vec3(0.0f, 1.0f, 0.0f));
        hmm_mat4 proj = HMM_Orthographic(0.0f, sapp_width(), 0.0f, sapp_height(), -1.0f, 5.0f);
        hmm_mat4 view_proj = HMM_MultiplyMat4(proj, view);

        sg_apply_bindings(&bindings);
        sg_apply_uniforms(SG_SHADERSTAGE_VS, 0, &view_proj, sizeof(hmm_mat4));
        sg_draw(0, 6, MAX_BUFFER_SIZE / sizeof(rectangle_t));
    }
    sg_end_pass();
    sg_commit();
}
```

**Comment by floatms** [comment]
Ah I see that makes sense, thanks for the snippets!
I actually called sg_apply_pipeline() inside the loop because I need to switch shaders inside the render loop of my actual program. So this was just a leftover from there.

Yeah, I'm also glad that this turned out not to be a bug but I hope that, as you said, there may be some value to extract from this.

Also, I took some performance numbers on the fixed version now and it seems like sgfx can draw 100 000 instanced semi-transparent rectangles (alpha value == 0.1) distributed across 400 buffers (also at random points on the screen with overlap) within 0.6 ms. Don't know if that is of interest for this issue but I found it pretty neat ;)

Cheers.
